### DISCUSSION TITLE
Feature Proposal: Always Random Wildcards

### DISCUSSION THREAD

**Discussion by DavidDragonsage** [lllyasviel/Fooocus#2761] [Ideas]
I really appreciate the new "Disable seed increment" option in the Developer Debug Mode Debug Tools section.

When used with arrays, it works very nicely to produce a variety of results based on the same seed.

But currently there is a problem in that "Disable seed increment" prevents wildcards from being chosen randomly.
 
So in this example prompt:
"portrait of an __ethnic__ woman"
(where the ethnic wildcard is a list of 330 worldwide ethnicities)
and an Image Number of 5, using the same seed and without incrementing that seed, Fooocus will produce five completely identical results, as shown in this extract from the console window:

[Fooocus] Processing prompts ...
[Wildcards] processing: portrait of an __ethnic__ woman
[Wildcards] portrait of an Ugandan woman
[Wildcards] processing: portrait of an __ethnic__ woman
[Wildcards] portrait of an Ugandan woman
[Wildcards] processing: portrait of an __ethnic__ woman
[Wildcards] portrait of an Ugandan woman
[Wildcards] processing: portrait of an __ethnic__ woman
[Wildcards] portrait of an Ugandan woman
[Wildcards] processing: portrait of an __ethnic__ woman
[Wildcards] portrait of an Ugandan woman
[Fooocus] Preparing Fooocus text #1 ...

With some minor changes to async_worker.py, wildcard random number generation is no longer affected by the "Disable seed increment" value, as shown:

[Fooocus] Processing prompts ...
[Wildcards] processing: portrait of an __ethnic__ woman
[Wildcards] portrait of an Ugandan woman
[Wildcards] processing: portrait of an __ethnic__ woman
[Wildcards] portrait of an Herzegovinian woman
[Wildcards] processing: portrait of an __ethnic__ woman
[Wildcards] portrait of an Corsican woman
[Wildcards] processing: portrait of an __ethnic__ woman
[Wildcards] portrait of an Sioux woman
[Wildcards] processing: portrait of an __ethnic__ woman
[Wildcards] portrait of an Assyrian woman
[Fooocus] Preparing Fooocus text #1 ...

So with this modification, Fooocus produces five different images from the same seed, allowing for a project theme to be developed using similar image composition or subject poses or perhaps an identical character that is doing different activities or is dressed differently based on changes in wildcard keywords.

I cannot imagine when you would want to run wildcards with a frozen seed because that defeats the purpose of wildcards. Subject to approval as usual, I would like to create a PR for this change.

**Comment by mashb1t** [comment]
The "Disable seed increment" option is very useful when used with the array prompt syntax introduced in PR https://github.com/lllyasviel/Fooocus/pull/1503 and is not intended to be used in combination with wildcards.
Please check out the PR for more examples and insights.

**Comment by DavidDragonsage** [comment]
Yes, as I mentioned at the top of this proposal, "Disable seed increment... When used with arrays, it works very nicely to produce a variety of results based on the same seed" and I do appreciate that.

This modification does not affect the operation of arrays. 

Again with the same seed as before, this time using an array to replicate what the wildcard produced in the last example, this is the test prompt:
"portrait of an [[Ugandan, Herzegovinian, Corsican, Sioux, Assyrian]] woman"

With an Image Number of 5, using the same seed, without incrementing that seed and with my modification in effect, the array exactly replicated the wildcard results I achieved previously, as shown in this extract from the console window:

Loading 1 new model
[Fooocus] Processing prompts ...
[Arrays] processing: portrait of an [[Ugandan, Herzegovinian, Corsican, Sioux, Assyrian]] woman
[Arrays] processing: portrait of an [[Ugandan, Herzegovinian, Corsican, Sioux, Assyrian]] woman
[Arrays] processing: portrait of an [[Ugandan, Herzegovinian, Corsican, Sioux, Assyrian]] woman
[Arrays] processing: portrait of an [[Ugandan, Herzegovinian, Corsican, Sioux, Assyrian]] woman
[Arrays] processing: portrait of an [[Ugandan, Herzegovinian, Corsican, Sioux, Assyrian]] woman
[Fooocus] Preparing Fooocus text #1 ...
[Prompt Expansion] portrait of an Ugandan woman, [snip]
[Fooocus] Preparing Fooocus text #2 ...
[Prompt Expansion] portrait of an Herzegovinian woman, [snip]
[Fooocus] Preparing Fooocus text #3 ...
[Prompt Expansion] portrait of an Corsican woman, [snip]
[Fooocus] Preparing Fooocus text #4 ...
[Prompt Expansion] portrait of an Sioux woman, [snip]
[Fooocus] Preparing Fooocus text #5 ...
[Prompt Expansion] portrait of an Assyrian woman, [snip]
[Fooocus] Encoding positive #1 ... 

I very often use both arrays and wildcards in my projects. There is no reason that "Disable seed increment" could not be just as useful for wildcards, for very similar reasons to its usefulness for arrays.

In this example, arbitrarily specifying five ethnicities in an array is at odds with my typical purpose of randomly diversifying images of people as much as possible. While Fooocus wildcard randomness is not ideal, it is a lot better than the randomness that I can come up with just out of my head!

In async_worker.py, we are looking at the addition of one line of code and the change of a variable name in the next line. Logically this will not affect the integrity of seed production, but I have tested to ensure that it does not.

The introduction of this small enhancement would enrich the Fooocus experience for a very small overhead, and would likely forestall future bewilderment as to why wildcards do not work in certain situations - precisely at the time when one would count on them to work.

**Comment by DavidDragonsage** [comment]
I have discovered that my modification slightly alters the Prompt Expansion added by Fooocus V2. This makes it impossible to exactly replicate a previously created image. So if we were to go ahead with this enhancement, I would likely need a bit of help - perhaps just a hint - in discovering why this is so. I have no idea at the moment how this modification could change the Prompt Expansion. 😞

**Comment by DavidDragonsage** [comment]
I slightly adjusted my coding and now the modification does not affect the Fooocus V2 Prompt Expansion. 🙂

**Comment by DavidDragonsage** [comment]
@mashb1t   I really hesitate to bring this back to your attention because I don't want to distract you from incorporating your branch into mainstream Fooocus!

There has been no support for this feature here but I did have someone track me down on Facebook to ask me for this patch. I actually don't use it myself because I much more interested in keeping my Fooocus completely current and not missing out on anything new! 😉

**Reply by mashb1t** [reply]
@DavidDragonsage I'm not certain if this is really something for the masses... Maybe you can create a PR so others can add your changes to their local installations or i can merge them e.g. with some additional developer debug checkbox support.

**Reply by DavidDragonsage** [reply]
@mashb1t  That's great! I would be happy to work up a PR for this. I do hope you can merge it into production Fooocus though, I don't think its harmful for anyone.

Personally I don't use any standalone PRs or even branches because its just too much effort to keep up with them all and also keep mainstream Fooocus current. And having multiple Fooocus's installed in separate folders is just not my idea of fun! 😉 

I agree, it doesn't seem to have wide support. On the other hand does it really need a debug checkbox? I can't see anyone actually wanting the wildcard to NOT increment when the seed is frozen. I can see a lot of people not caring about whether it does or doesn't, but I can't imagine why someone would ever want a wildcard to be frozen. If you want the keyword(s) to always be the same, you just type them into the prompt - or use a wildcard with only one entry.

**Comment by DavidDragonsage** [comment]
I have added a PR to my fork. I cannot remember how to bring that into the main PRs.
